---
title: "Differential_expression_analysis_Final_submission"
author: "Vikas Kaushik"
date: "`r Sys.Date()`"
output:
  pdf_document: default
  html_document: default
---
Load relevant R packages.

```{r, message=FALSE, warning=FALSE}
# Load the ballgown package for RNA-seq analysis
library(ballgown)

# Load the tidyverse package for data manipulation and visualization
library(tidyverse)

# Load the ggplot2 package for creating versatile plots
library(ggplot2)

# Load RSkittleBrewer for additional color palettes
library(RSkittleBrewer)

# Load genefilter for filtering genes based on certain criteria
library(genefilter)

# Load dplyr for data manipulation
library(dplyr)

# Load devtools for development-related functions
library(devtools)

# Load cowplot for combining and arranging plots
library(cowplot)

# Load DESeq2 for differential gene expression analysis
library(DESeq2)

# Load EnhancedVolcano for creating volcano plots
library(EnhancedVolcano)

# Load pheatmap for creating heatmaps
library(pheatmap)

```

Load the phenotype data for the samples stored in  geuvadis_phenodata.csv

```{r, message=FALSE, warning=FALSE}
# Make sure to chnage the path to 'D:/NGS/chrX_data' where file is located
setwd('D:/NGS/chrX_data')

# Read the phenotype data into an object
geuvadis_phenodata <- read.csv("geuvadis_phenodata.csv")

# View the data frame
class(geuvadis_phenodata)

```
Read in the expression data that were calculated by StringTie. 

```{r, message=FALSE, warning=FALSE}
# Make sure geuvadis_phenodata is a valid data frame, Convert it to a data frame if needed
geuvadis_phenodata <- as.data.frame(geuvadis_phenodata)

# Create ballgown object
bg_chrX = ballgown(dataDir = "ballgown", samplePattern = "ERR", pData = geuvadis_phenodata)

```

Filter to remove low-abundance genes. 

```{r, warning=FALSE}
# Filter genes in bg_chrX with row-wise variance greater than 1
bg_chrX_filt <- subset(bg_chrX, "rowVars(texpr(bg_chrX)) > 1", genomesubset = TRUE)

```

Identify transcripts that show statistically significant differences between groups. 

```{r}
# DEG analysis on transcripts, testing for the effect of 'sex' with adjustment for 'population'
results_transcripts = stattest(bg_chrX_filt, feature = "transcript", covariate = "sex", adjustvars = c("population"), getFC = TRUE, meas = "FPKM")


```

Identify genes that show statistically significant differences between groups.

```{r}
# DEG analysis on genes, testing for the effect of 'sex' with adjustment for 'population'
results_genes = stattest(bg_chrX_filt, feature = "gene", covariate = "sex", adjustvars = c("population"), getFC = TRUE, meas = "FPKM")

```

Add gene names and gene IDs to the results_transcripts data frame.

```{r}
# Combine results of transcript-level differential expression analysis with gene names and IDs
results_transcripts = data.frame(
  geneNames = ballgown::geneNames(bg_chrX_filt),
  geneIDs = ballgown::geneIDs(bg_chrX_filt),
  results_transcripts
)
```

Sort the results from the smallest P value to the largest.

```{r}
# Arrange results_transcripts and results_genes data frames based on p-value
results_transcripts = arrange(results_transcripts, pval)
results_genes = arrange(results_genes, pval)

```

Write the results to csv files.

```{r}
# Write transcript-level results to a CSV file
write.csv(results_transcripts, "chrX_transcript_results.csv", row.names = FALSE)

# Write gene-level results to a CSV file
write.csv(results_genes, "chrX_gene_results.csv", row.names = FALSE)

```

Identify transcripts and genes with a significant level of expression

```{r}
# Subset transcript-level results with p-value < 0.05
subset_transcripts = subset(results_transcripts, pval < 0.05)

# Subset gene-level results with p-value < 0.05
subset_genes = subset(results_genes, pval < 0.05)

# Count the number of up- and down-regulated genes
upregulated <- sum(subset_genes$fc > 0)
downregulated <- sum(subset_genes$fc < 0)

# Number of up-regulated genes
cat("Number of upregulated genes:", upregulated, "\n")

## Number of down regulated genes
cat("Number of downregulated genes:", downregulated, "\n")

```
Adds touch of tropical flair to our visualizations, making them visually appealing and thematic

```{r}
# Define a tropical color palette
tropical = c('darkorange', 'dodgerblue', 'hotpink', 'limegreen', 'yellow')

# Set the color palette for plots
palette(tropical)

```

Show the distribution of gene abundances (measured as FPKM values) across samples, colored by sex.

```{r, fig.cap="Distribution of FPKM values across the 12 samples"}
# Extract FPKM values from ballgown object
fpkm = texpr(bg_chrX, meas = "FPKM")

# Log2 transformation of FPKM values
fpkm = log2(fpkm + 1)

geuvadis_phenodata$sex = as.factor(geuvadis_phenodata$sex)

# Create a boxplot with colors representing different sexes
Figure_1 <- boxplot(fpkm, col = geuvadis_phenodata$sex, las = 2, ylab = 'log2(FPKM+1)')

# Add title
title(main = "Distribution of FPKM values across the 12 samples")

# Initialize the PNG device
png("Distribution of FPKM values across the 12 samples.png")

# Save the base R boxplot as a figure
dev.copy(png, "Distribution of FPKM values across the 12 samples.png")

dev.off()

```

```{r, fig.cap="Distribution of Gene Abundances"}
# Create a PNG file with specified dimensions
png(file = "Gene Abundance.png", width = 600, height = 300)

# Create a data frame with 'sex' and 'fpkm' columns
plot_data <- data.frame(sex = geuvadis_phenodata$sex, fpkm = fpkm[12,])

# Create a new column named "Sample_id" with the values from the row names from plot_data
plot_data$Sample_id <- rownames(plot_data)

# Remove "FPKM." from the "Sample_id" column using sub function
plot_data$Sample_id <- sub("FPKM\\.", "", plot_data$Sample_id)

# Remove the original row names
rownames(plot_data) <- NULL

# Create a boxplot with colors representing different sexes
ggplot(plot_data, aes(x = sex, y = fpkm, fill = sex)) +
  geom_boxplot() +
  labs(title = "Distribution of Gene Abundances",
       x = "Sex",
       y = "FPKM Values") +
  scale_fill_manual(values = c('darkorange', 'dodgerblue')) +
  
  # Scale x-axis 
  scale_x_discrete(name = "Gender",
                   labels = c("Female", "Male")) +
  
  # Scale y-axis
  scale_y_continuous(name = "Fragments Per Kilobase Million",
                     breaks = seq(0, 5, by = 0.2))+
  
  # Remove legend to create a cleaner appearance
  theme(legend.position = "none")

```

Visualization of the assembled transcripts is done with the plotTranscripts function. Transcripts aew colored by expression level. This plot colors transcripts by expression level

```{r, fig.cap="Transcript Visualization by Expression Levels"}
# Set larger plot size
png("Transcript_plot.png", width = 1200, height = 600)

# Plot transcripts for a specific gene in all samples
gene_index <- 1729
gene_id <- ballgown::geneIDs(bg_chrX)[gene_index]

# Plot transcripts for two samples
plotTranscripts(gene_id, bg_chrX, 
  samples = c('ERR188044', 'ERR188104', 'ERR188454', 'ERR188257', 'ERR188383', 'ERR188401','ERR188234', 'ERR188245', 'ERR188273', 'ERR188337', 'ERR188428', 'ERR204916'),
  colorby = "transcript", meas = "FPKM", legend = TRUE, labelTranscripts = FALSE,
  main = paste("Gene ", gene_id, " Expression across the different Samples"),
  blackBorders = TRUE, log = FALSE, logbase = 2, customCol = NULL, customOrder = NULL)

# Reset plot parameters to default and close the device
dev.off()

#open the image
system("open Transcript_plot.png")

```
Group similar transcripts together

```{r}

# group a geneâ€™s assembled transcripts into clusters
clusterTranscripts(gene_id, bg_chrX, k = 2, method = c("kmeans"))

#visualize the transcript clusters:
plotLatentTranscripts(gene_id, bg_chrX, k=2, method='kmeans', returncluster=FALSE)

```

Plot the average expression levels for all transcripts of a gene within different groups using the plotMeans function. 

```{r}
# Define the gene whose transcripts will be plotted.
Gene_id <- 'MSTRG.56'

# Create a plot of means grouped by 'sex'
plotMeans(Gene_id, bg_chrX_filt, groupvar = "sex", colorby = c("transcript"), legend = FALSE)

```

```{r, fig.cap="Visualizing Transcriptomic Changes with the Volcano Plot"}
# Create an Volcano plot to visualize statistically significant difference in expression levels
EnhancedVolcano(
  results_genes,
  lab = rownames(results_genes),
  x = 'fc',
  y = 'pval',
  xlim = c(-4, 4),
  ylim = (c(0, 10)),
  title = "Volcano Plot (p-value < 0.05)")
  
```
```{r, fig.cap="Visualizing upregulated genes with Volcano Plot"}
# Adjust the threshold values as needed
threshold_logFC <- 0.0
threshold_pvalue <- 0.05

# Create the volcano plot to visualize most relevant genes for further investigation.
plot(results_genes$fc, -log10(results_genes$pval), 
     xlab = "log2(Fold Change)", 
     ylab = "-log10(p-value)",
     main = "Volcano Plot")

# Add a threshold line for log-fold change
abline(v = threshold_logFC, col = "red", lty = 2)

# Add a threshold line for p-value
abline(h = -log10(threshold_pvalue), col = "blue", lty = 2)

```

```{r, fig.cap="A Heatmap Exploration of the Top 20 Genes Across 12 Samples with gene ID"}
# Assuming bg_chrX_filt is your Ballgown object
expression_matrix <- as.matrix(texpr(bg_chrX_filt))

# Get the row names of the top N genes based on maximum expression values
top_genes_ids <- names(sort(apply(expression_matrix, 1, max), decreasing = TRUE)[1:10])

# Extract expression data for the top N genes
top_expression <- expression_matrix[top_genes_ids, ]

# Create a heatmap using the numeric expression matrix of top genes with gene ID
heatmap(top_expression, 
        Colv = NA,    # To disable hierarchical clustering of columns
        Rowv = NA,    # To disable hierarchical clustering of rows
        scale = "row", # Scale rows (genes)
        main = paste("Highly Expressed Genes Heatmap")
)
```

```{r, fig.cap="A Heatmap Exploration of the Top 20 Genes Across 12 Samples with Gene name"}
# Extract gene names corresponding to the top gene IDs
matching_gene_names <- results_transcripts$geneIDs[results_transcripts$id %in% top_genes_ids]

# Extract expression data for the specified genes
expression_data_to_plot <- expression_matrix[top_genes_ids, ]

# Create a heatmap using the numeric expression matrix of specified genes with gene names
heatmap(expression_data_to_plot, 
        Colv = NA,    # To disable hierarchical clustering of columns
        Rowv = NA,    # To disable hierarchical clustering of rows
        scale = "row", # Scale rows (genes)
        main = "Heatmap for higly expressed genes",
        labRow = matching_gene_names  # Use matching gene names as row labels
)

```

```{r, fig.cap="A Heatmap Exploration of the Top 20 Genes Across 12 Samples"}

# Get the row names of the top N genes based on maximum expression values
top_genes_ids <- names(sort(apply(expression_matrix, 1, max), decreasing = TRUE)[1:20])

# Extract gene names corresponding to the top gene IDs
matching_gene_names <- results_transcripts$geneIDs[results_transcripts$id %in% top_genes_ids]

# Extract expression data for the specified genes
expression_data_to_plot <- expression_matrix[top_genes_ids, ]

# Expand 'anno_color' to match the number of columns in 'heatmap_data'
anno_color_expanded <- data.frame(Annotation = matching_gene_names)

# Set up colors for the annotations
colors <- c("White", "white")

# Creating a heatmap with the expanded 'anno_color' and specified colors
pheatmap(expression_data_to_plot,
         annotation_col = anno_color_expanded,
         annotation_colors = list(Annotation = colors),
         scale = "row",
         main = "Heatmap for higly expressed genes",
         cutree_rows = 2,
         cutree_cols = 3)

```

